### DISCUSSION TITLE
Custom tls.Config and dialFunc in go-ora

### DISCUSSION THREAD

**Discussion by baldeephira** [sijms/go-ora#497] [Q&A]
Is there a way to set custom tls.Config and dial function for go-ora driver?  Looking for something like what is possible with postgresql golang driver (pgx), shown below.

I saw that tls.Config is set in the session inside negotiate() function, need a way to set it externally.

```
import "github.com/jackc/pgx/v5"
import "github.com/jackc/pgx/v5/stdlib"

... 

cfg, _ := pgx.ParseConfig(myDSN)

// set my custom tls.Config
cfg.TLSConfig = myCustomTlsCfg

// set my custom dial function
cfg.DialFunc = func(ctx context.Context, network, addr string) (net.Conn, error) {
	// my custom dial handler
}

dsn := stdlib.RegisterConnConfig(cfg)
db, err := sql.Open("pgx", dsn)
```

**Comment by sijms** [comment]
ok I will try to add it later

**Reply by baldeephira** [reply]
Thanks Sammy!  Let me know if I can assist in anyway.

**Comment by baldeephira** [comment]
Hi Samy, any update on this?  Thanks!

**Comment by sijms** [comment]
sorry for late I will try to finish it in this release

**Comment by baldeephira** [comment]
Thank you so much, much appreciated.  Pls let me know if I can help in anyway?  Best wishes!On Apr 17, 2024, at 8:44 AM, Samy Sultan ***@***.***> wrote:﻿
sorry for late I will try to finish it in this release

**Comment by sijms** [comment]
I add this feature in v2.8.13 you can test it and write your feedback

**Reply by baldeephira** [reply]
I want to test 2 different connection configuration scenarios:
1. use "ssh dialer" to reach oracle db instance running behind firewall using ssh tunneling
2. use mTLS with self-signed certificates to talk to oracle db instance

I see a way to accomplish scenario (1) with following code:
```
// init custom ssh dialer
d := sshDialer{sshClient: my_ssh_client}

config, err := go_ora.ParseConfig(my_dsn)
config.Dialer = d
go_ora.RegisterConnConfig(config)
db, err := sql.Open("oracle", "")
```

However, I'm not sure how to achieve/test scenario (2).  Is this possible via Wallet configuration?  
I want to achieve the following:
```
// init/load certificates & private keys from external KMS
var root_CA_cert []byte
var client_cert []byte
var client_private_key []byte

// init root CA certs
rootCertPool := x509.NewCertPool()
rootCertPool.AppendCertsFromPEM(root_CA_cert)

// for mTLS, add client certificate
cert, _ := tls.X509KeyPair([]byte(client_cert), []byte(client_key))
certs := make([]tls.Certificate, 1)
certs[0] = cert

tlsCfg = &tls.Config{
	RootCAs:            rootCertPool,
	Certificates:       certs,
	InsecureSkipVerify: true, // needed for self signed certs
}
```

**Comment by sijms** [comment]
you can use it under the wallet which is member of connection configuration
```golang
type Wallet struct {
	file                string
	password            []byte
	salt                []byte
	sha1Iteration       int
	algType             int
	credentials         []WalletCredential
	Certificates        [][]byte
	PrivateKeys         [][]byte
	CertificateRequests [][]byte
}
```

**Comment by baldeephira** [comment]
Verified with v2.8.13 and it works.  Thanks Samy!

1. Successfully tested connection to Oracle Autonomous database using mTLS.  The private key and certificate chain was loaded using externalized PEM strings as read from eWallet.pem.  Note: the private key in eWallet.pem uses PKCS8 format, need to convert this to PKCS1 (RSA PRIVATE KEY) format and then load the []byte.
2. Successfully tested connection to Oracle Autonomous database via ssh tunnel.  The sshClient instance was configured and set in Wallet instance
